framework/src/timesteppers/FixedPointIterationAdaptiveDT.C

```diff
@@ -0,0 +1,112 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FixedPointIterationAdaptiveDT.h"
+#include "FEProblemBase.h"
+#include "Transient.h"
+
+registerMooseObject("MooseApp", FixedPointIterationAdaptiveDT);
+
+InputParameters
+FixedPointIterationAdaptiveDT::validParams()
+{
+  InputParameters params = TimeStepper::validParams();
+  params.addClassDescription(
+      "Computes time step size based on a target number of fixed point iterations");
+  params.addRequiredParam<Real>("dt_initial", "The initial time step size");
+  params.addRequiredParam<unsigned int>("target_iterations",
+                                        "The target number of fixed point iterations");
+  params.addRequiredParam<unsigned int>(
+      "target_window",
+      "The number of iterations added to and subtracted from 'target_iterations' to determine the "
+      "iteration window; the time step size will increase if the iterations were below "
+      "'target_iterations' - 'target_window' and will decrease if the iterations were above "
+      "'target_iterations' + 'target_window'.");
+  params.addParam<Real>("increase_factor",
+                        1.2,
+                        "Factor by which the previous time step size will increase if the previous "
+                        "number of fixed point iterations was below the target window minimum "
+                        "('target_iterations' - 'target_window').");
+  params.addParam<Real>("decrease_factor",
+                        0.8,
+                        "Factor by which the previous time step size will decrease if the previous "
+                        "number of fixed point iterations was above the target window maximum "
+                        "('target_iterations' + 'target_window').");
+
+  return params;
+}
+
+FixedPointIterationAdaptiveDT::FixedPointIterationAdaptiveDT(const InputParameters & parameters)
+  : TimeStepper(parameters),
+    _dt_initial(getParam<Real>("dt_initial")),
+    _target_center(getParam<unsigned int>("target_iterations")),
+    _target_window(getParam<unsigned int>("target_window")),
+    _target_min(_target_center - _target_window),
+    _target_max(_target_center + _target_window),
+    _increase_factor(getParam<Real>("increase_factor")),
+    _decrease_factor(getParam<Real>("decrease_factor")),
+    _dt_old(declareRestartableData<Real>("dt_old", 0.0)),
+    _fp_its(declareRestartableData<unsigned int>("fp_its", 0))
+{
+}
+
+void
+FixedPointIterationAdaptiveDT::init()
+{
+  TimeStepper::init();
+
+  if (!_fe_problem.hasMultiApps())
+    mooseError("This time stepper can only be used if there are MultiApps in the problem.");
```

Comment:
need to check that fixed point iterations are in use, which the executioner would know about